designs/2019-typescript-eslintrc-support/README.md

```diff
@@ -0,0 +1,351 @@
+- Start Date: 2019-11-18
+- RFC PR: https://github.com/eslint/rfcs/pull/50
+- Authors: Gareth Jones ([@g-rath](https://github.com/g-rath))
+
+# First-Class Support For `.eslintrc.ts`
+
+## Summary
+
+Support the ability to consume `.eslintrc.ts` configuration files natively.
+
+## Motivation
+
+Currently configuration settings are validated by JSON Schema, which while
+powerful does not have ubiquitous & high quality support across editors &
+tooling.
+
+It also tends to generate somewhat cryptic messages, especially when using
+complex elements such as combining `oneOf`, `anyOf`, and other such schemas.
+
+Meanwhile, TypeScript is a powerful & popular superset of Javascript that has
+high quality support as a first class language in the majority of editors, and
+is quickly gaining ubiquitous support in general.
+
+By allowing first-class support for `.eslintrc.ts`, the amount of configuration
+problems & general confusion over such experienced by users can be reduced,
+since it would enable us to provide tooling to type configuration files and so
+allow TypeScript to catch ESLint configuration errors at compile time, rather
```

Comment:
Wouldn’t compile time be exactly the same as run time in this case? Since ESLint would execute `ts-node` during runtime and thus compiling `.eslintrc.ts` at runtime.

Reply:
Yes, when running ESLint, but the act of "compiling" is done by tsc, meaning you can check the `.eslintrc` w/o running ESLint.

This is particularly useful for editors, who do run the TypeScript server separately to ESLint; one advantage to this in particular is that you'll get squiggles for *all* the errors in the configuration at once, since TypeScript will check the whole file every time (i.e it won't bail out if a rule name is invalid).

Reply:
In addition , this would also allow you to validate the configuration without ESLint nor the plugins, since you only require the type information.

This would be particularly valuable for configuration repos, where they might configurations for optional packages that don't make to sense to have installed themselves.

An immediate example that springs to mind is `eslint-config-prettier`: they have a configuration file for `@typescript-eslint` - in order to validate that the settings are correct, you would have to install `@typescript-eslint/eslint-plugin`, which is "big" given that there is no use for it outside of that, and it would probably require additional work on top of that, since unless you installed TypeScript (an 80mb package) & `@typescript-eslint/parser`, you'd have to handle `eslint` erroring since `@typescript-eslint/eslint-plugin` would complain about not having TypeScript for it's type-requiring rules.

Meanwhile, this proposal would mean you could just you'd just be able to install `@typescript-eslint/eslint-plugin-types` which would be drastically smaller, and use TypeScript to valid the configuration.

Meanwhile, `@typescript-eslint` themselves can just have `@typescript-eslint/eslint-plugin-types` as a dependency of `eslint-plugin` to avoid duplication both in their source and their published package.

----

Keep in mind that the above example extends beyond the `@typescript-eslint/eslint-plugin` - just b/c it's easy to think it's self defeating by requiring you to switch to TypeScript in the first place, but this also applies to other packages & parsers.

Reply:
In light of my above comment, I'm going to mark this as resolved for now. Let me know if you'd like this example added to the actual RFC :)